File: 백준/Gold/16936. 나3곱2/나3곱2.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const N = Number(input.shift());
const numArr = input[0].split(' ').map(BigInt);
const visited = new Array(N).fill(false);
let done = false;

function dfs(stack, now) {
    if (stack.length > 1) {
        const f = stack[now-2];
        const b = stack[now-1];
        if((f % 3n !== 0n || f / 3n !== b) && f * 2n !== b) return;
    }
    if (now >= N) {
        if (!done) {
            console.log(stack.join(' '));
            done = true;
        }
        return;
    }

    for (let i = 0; i < N; i++) {
        if (!visited[i]) {
            visited[i] = true;
            stack.push(numArr[i]);
            dfs(stack, now + 1);
            stack.pop();
            visited[i] = false;
        }
    }
}

dfs([], 0);

File: 백준/Gold/16936. 나3곱2/나3곱2.py
from itertools import permutations

n = int(input())
arr = list(map(int, input().split()))
visited = [False] * n

def dfs(temp, depth):
    if depth >= 2:
        a = temp[depth-2]
        b = temp[depth-1]

        if b != a * 2 and not (a % 3 == 0 and b == a // 3):
            return

    if depth == n:
        print(" ".join(map(str,temp)))
        return

    for i in range(n):
        if not visited[i]:
            visited[i] = True
            temp.append(arr[i])
            dfs(temp, depth+1)
            visited[i] = False
            temp.pop()

dfs([], 0)